### ISSUE TITLE
[Bug]: history.push no longer triggers a re-render when a search param has changed

### ISSUE THREAD

**Issue by mattp94** [ReactTraining/react-router#7918]
### What version of React Router are you using?

5.2.0

### Steps to Reproduce

Hello 👋, I observed a strange behavior after upgrading `react-router-dom` from **v5.1.2** to **v5.2.0**. It looks like `history.push` or `history.replace` (taken from `useHistory`) doesn't trigger anymore a re-render when a search param has changed.

### Expected Behavior

Here is a simple example with **v5.1.2**:
https://codesandbox.io/s/with-react-router-dom-512-before-zljp8

As you can see, calling `history.push` changes the URL as well as the counter value. So everything works well!

### Actual Behavior

Now, when I test with **v5.2.0**, it no longer works:
https://codesandbox.io/s/with-react-router-dom-520-after-z796t

Here, calling `history.push` changes the URL but doesn't re-render the counter... It's a bit annoying as this minor version has introduced a breaking change in our application.

Is it normal? 🙂

**Comment by renatobenks** [comment]
it might be related to https://github.com/ReactTraining/react-router/pull/7103 or https://github.com/ReactTraining/react-router/pull/7288, but much probably was the update of mini-create-react-context as the changes on `useLocation` don't seem to have any problems

**Comment by timdorr** [comment]
You're not subscribing to any location change. You're relying on an old performance-related bug that we fixed in 5.2.  Use the `useLocation` hook instead and it works just fine:

```js
export default function App() {
  const history = useHistory();
  const location = useLocation();

  const searchParams = new URLSearchParams(location.search);
  const counter = Number(searchParams.get("counter"));

  const handleClick = () => {
    searchParams.set("counter", counter + 1);
    history.push("?" + searchParams);
  };

  return (
    <>
      <button onClick={handleClick}>Increment</button>
      {counter}
    </>
  );
}
```

**Comment by mattp94** [comment]
Thanks a lot @timdorr 👍
